Test/Unit/Client.php

```diff
@@ -0,0 +1,820 @@
+                    $self
+                        ->boolean($_enable)
+                            ->isTrue()
+                        ->integer($_type)
+                            ->isEqualTo($socket::ENCRYPTION_TLS)
+                        ->variable($_sessionStream)
+                            ->isNull();
+
+                    return true;
+                },
+                $this->calling($socket)->setStreamBlocking = function ($_block) use ($self) {
+                    $self
+                        ->boolean($_block)
+                            ->isTrue();
+
+                    return true;
+                },
+                $this->calling($socket)->writeAll = function ($_data) use ($self, $endPoint, $host, $challenge) {
+                    $self
+                        ->string($_data)
+                            ->isEqualTo(
+                                'GET ' . $endPoint . ' HTTP/1.1' . CRLF .
+                                'Host: ' . $host . CRLF .
+                                'User-Agent: Hoa' . CRLF .
+                                'Upgrade: WebSocket' . CRLF .
+                                'Connection: Upgrade' . CRLF .
+                                'Pragma: no-cache' . CRLF .
+                                'Cache-Control: no-cache' . CRLF .
+                                'Sec-WebSocket-Key: ' . $challenge . CRLF .
+                                'Sec-WebSocket-Version: 13' . CRLF . CRLF
+                            );
+
+                    return strlen($_data);
+                },
+                $this->calling($socket)->read = function ($_size) use ($self, $challenge, $response) {
+                    $self
+                        ->integer($_size)
+                            ->isEqualTo(2048);
+
+                    return $response;
+                },
+                $this->calling($client)->getNewChallenge = $challenge
+            )
+            ->exception(function () use ($client) {
+                $this->invoke($client)->doHandshake();
+            })
+                ->isInstanceOf(Websocket\Exception\BadProtocol::class);
+    }
+
+    public function case_get_new_challenge()
+    {
+        $this
+            ->given(
+                $this->mockGenerator->orphanize('__construct'),
+                $client = new \Mock\Hoa\Websocket\Client()
+            )
+            ->when(function () use ($client) {
+                for ($i = 0; $i < 1000; ++$i) {
+                    $this
+                        ->string($client->getNewChallenge())
+                            ->matches('/^[A-Za-z0-9]{21}[AQgw]==$/');
+                }
+            });
+    }
+
+    public function case_close()
+    {
+        return $this->_case_close(true);
+    }
+
+    public function case_close_with_no_protocol_implementation()
+    {
+        return $this->_case_close(null);
```

Comment:
Yup but data provider counts only 1 test case, and I prefer explicit naming too.